src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessorWorkflow.java

```diff
@@ -67,22 +75,71 @@ public void execute(
     private List<CompoundTopDocs> getQueryTopDocs(final List<QuerySearchResult> querySearchResults) {
         List<CompoundTopDocs> queryTopDocs = querySearchResults.stream()
             .filter(searchResult -> Objects.nonNull(searchResult.topDocs()))
-            .filter(searchResult -> searchResult.topDocs().topDocs instanceof CompoundTopDocs)
-            .map(searchResult -> (CompoundTopDocs) searchResult.topDocs().topDocs)
+            .map(querySearchResult -> querySearchResult.topDocs().topDocs)
+            .map(CompoundTopDocs::create)
             .collect(Collectors.toList());
+        if (queryTopDocs.size() != querySearchResults.size()) {
+            log.debug("Some of querySearchResults are not produced by hybrid query");
+        }
         return queryTopDocs;
     }
 
-    private void updateOriginalQueryResults(final List<QuerySearchResult> querySearchResults, final List<CompoundTopDocs> queryTopDocs) {
-        for (int i = 0; i < querySearchResults.size(); i++) {
-            QuerySearchResult querySearchResult = querySearchResults.get(i);
-            if (!(querySearchResult.topDocs().topDocs instanceof CompoundTopDocs) || Objects.isNull(queryTopDocs.get(i))) {
-                continue;
-            }
-            CompoundTopDocs updatedTopDocs = queryTopDocs.get(i);
-            float maxScore = updatedTopDocs.totalHits.value > 0 ? updatedTopDocs.scoreDocs[0].score : 0.0f;
-            TopDocsAndMaxScore updatedTopDocsAndMaxScore = new TopDocsAndMaxScore(updatedTopDocs, maxScore);
+    private void updateOriginalQueryResults(
+        final List<QuerySearchResult> querySearchResults,
+        final FetchSearchResult fetchSearchResult,
+        final List<CompoundTopDocs> queryTopDocs,
+        final boolean isSingleShard
+    ) {
+        int queryTopDocsIndex = 0;
+        for (QuerySearchResult querySearchResult : querySearchResults) {
+            CompoundTopDocs updatedTopDocs = queryTopDocs.get(queryTopDocsIndex++);
+            float maxScore = updatedTopDocs.getTotalHits().value > 0 ? updatedTopDocs.getScoreDocs()[0].score : 0.0f;
+
+            // create final version of top docs with all updated values
+            TopDocs topDocs = new TopDocs(updatedTopDocs.getTotalHits(), updatedTopDocs.getScoreDocs());
+
+            TopDocsAndMaxScore updatedTopDocsAndMaxScore = new TopDocsAndMaxScore(topDocs, maxScore);
             querySearchResult.topDocs(updatedTopDocsAndMaxScore, null);
         }
+        // a workaround for a single shard case, fetch has happened, and we need to update both fetch and
+        // query results
+        if (isSingleShard && querySearchResults.size() == 1) {
```

Comment:
agree on splitting into two methods for fetch and query results.

as for the general idea - that logic is there in core from the beginning, it doesn't change frequently. We cannot do one general flow of processing because for everything other than shard == 1 there will be empty fetch results. Normal sequence is QUERY phase - call processors registered between QUERY and FETCH - FETCH phase. Only for one shard this is different. You can think of our approach as a workaround for that behavior in core.

Reply:
I see what you mean, let me check if we can rely on fetch results being present / not null / not empty. I've added few integ tests for a single shard case, it should give some conference

Reply:
changed check to "if fetch results present"